### PR TITLE
Extract verified session logic into a concern

### PR DESCRIPTION
**Pull request by segiddins**
It was being duplicated a lot!

### CODE REVIEW

File: app/controllers/api_keys_controller.rb [hunk 1]
```diff
@@ -1,9 +1,8 @@
 class ApiKeysController < ApplicationController
   include ApiKeyable
-  before_action :redirect_to_signin, unless: :signed_in?
-  before_action :redirect_to_new_mfa, if: :mfa_required_not_yet_enabled?
-  before_action :redirect_to_settings_strong_mfa_required, if: :mfa_required_weak_level_enabled?
-  before_action :redirect_to_verify, unless: :password_session_active?
+
+  include SessionVerifiable
+  verify_session_before
```
Comment (martinemde) [inline app/controllers/api_keys_controller.rb:5]: The class method obfuscates more than it simplifies, imho. A single before action method with the conditions as regular ruby code would be canonical, easier to read and easy find. I'd just drop it in ApplicationController so controllers that need a custom redirect could override `redirect_to_verify` and call super.

**Review by martinemde — COMMENTED** [review]
I'm worried the redirects are going to cause problems. I added an idea for how you might approach this differently to make it easier to handle difference between controllers.

File: app/controllers/concerns/session_verifiable.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      before_action :redirect_to_settings_strong_mfa_required, **opts, if: :mfa_required_weak_level_enabled?
+      before_action :redirect_to_verify, **opts, unless: :password_session_active?
+    end
+  end
+
+  included do
+    private
+
+    def redirect_to_verify
+      redirect_uri = request.path_info
```
[the remark this answers is not in the document]
Comment (segiddins) [reply app/controllers/concerns/session_verifiable.rb:17]: I don't think there are any post-only URIs?

File: app/controllers/owners_controller.rb [hunk 3]
```diff
@@ -1,7 +1,9 @@
 class OwnersController < ApplicationController
+  include SessionVerifiable
+
   before_action :find_rubygem, except: :confirm
   before_action :render_forbidden, unless: :owner?, except: %i[confirm resend_confirmation]
-  before_action :redirect_to_verify, unless: :password_session_active?, only: %i[index create destroy]
+  verify_session_before only: %i[index create destroy]
```
[the remark this answers is not in the document]
Comment (segiddins) [reply app/controllers/owners_controller.rb:6]: agreed, was just dropping it in where the prior `redirect_to_verify` was

File: app/controllers/concerns/session_verifiable.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinemde) [reply app/controllers/concerns/session_verifiable.rb:17]: resend_confirmation and confirm?

File: app/controllers/api_keys_controller.rb (hunk 1 above)
Comment (segiddins) [reply app/controllers/api_keys_controller.rb:5]: the issue with collapsing to a single method is you lose the callback chain being halted as soon as one of them `redirect`s, meaning you have to put those conditionals into the method

Comment (martinemde) [reply app/controllers/api_keys_controller.rb:5]: That doesn't seem like a big problem to me. Mostly I just removed commas and colons from the above code, then added the returns.

```ruby
   before_action :verify_session

   # ...

    def verify_session
      return redirect_to_signin unless signed_in?
      return redirect_to_new_mfa if mfa_required_not_yet_enabled?
      return redirect_to_settings_strong_mfa_required if mfa_required_weak_level_enabled?
      return redirect_to_verify unless password_session_active?
    end
```

The callback code builds a whole stack and is certainly slower than this one method, plus then we save a module and we don't need to go look up `verify_session_before` to see that it's actually just a set of before filters`.

Comment (martinemde) [reply app/controllers/api_keys_controller.rb:5]: Oh, I looked closer at the filters themselves and they have their own conditions inside that might avoid the redirect. Unless you wanted to refactor that, then I agree with your original version for now.

**Review by martinemde — APPROVED** [review]
Based on digging further into the `redirect_to_*` methods, I agree with your original approach.

File: app/controllers/owners_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (segiddins) [reply app/controllers/owners_controller.rb]: looks like changing it broke tests :(

**Pull request merged**